### PR TITLE
Proof of concept optimization when both rows and cols are unspecified

### PR DESCRIPTION
**Pull request by yaccos**
From the definitions of `R_INDEX_OP` and `R_INDEX_GET`, I see that these macros check both arguments for NA-values. In the cases we know that none of the operands are NA, we can avoid wasting CPU cycles on this checking. In this case, I have modified `rowSums2()` to skip checking for index NA values when neither row nor col subsets are provided. From my own tests, I have registered a performance boost and would like you to try it out as well. If implemented consistently, we could probably mitigate most of issue #212.

### CODE REVIEW

**Comment by HenrikBengtsson** [comment]
Good points.  To clarify for others, and to make sure I got this correct, in matrixStats (<= 0.60.0), we had:

https://github.com/HenrikBengtsson/matrixStats/blob/be683978aa3c6bd8866c8e6e48aa424b47554d19/src/000.templates-types.h#L371-L377

whereas in matrixStats (>= 0.60.1), we now have:

https://github.com/HenrikBengtsson/matrixStats/blob/0158d9a34c5485cb4c7b55c164ee30a6b31bb9a2/src/000.templates-types.h#L151-L152

In other words, before the code did _not_ check for missing values when not subsetting by `rows` or `cols`, whereas now, `R_INDEX_OP` and `R_INDEX_GET` always checks for missing values.

So, yes, I think this explains the performance degradation going from matrixStats 0.60.0 to 0.60.1.

File: src/rowSums2_lowlevel_template.h [hunk 1]
```diff
@@ -50,21 +50,33 @@ void CONCAT_MACROS(rowSums2, X_C_SIGNATURE)(X_C_TYPE *x, R_xlen_t nrow, R_xlen_t
...
 
     sum = 0.0;
 
     for (jj=0; jj < ncols; jj++) {
-      if (nocols) {
-        if (byrow) idx = R_INDEX_OP(rowIdx, +, jj*nrow);
-        else idx = R_INDEX_OP(rowIdx, +, jj);
-      } else {
-        idx = R_INDEX_OP(rowIdx, +, colOffset[jj]);
+      if (norows && nocols){
```
Comment (HenrikBengtsson) [inline src/rowSums2_lowlevel_template.h:62]: Please add missing whitespace

File: src/rowSums2_lowlevel_template.h [hunk 2]
```diff
@@ -50,21 +50,33 @@ void CONCAT_MACROS(rowSums2, X_C_SIGNATURE)(X_C_TYPE *x, R_xlen_t nrow, R_xlen_t
...
+        /*
+         * In this special case, we can eliminate
+         * the possibility of having NA indicies
+         */
+        if (byrow) idx = rowIdx + jj*nrow;
+        else idx = rowIdx + jj;
+        value = x[idx];
+      }
+      else {
```
Comment (HenrikBengtsson) [inline src/rowSums2_lowlevel_template.h:71]: Please use `} else {` as we use elsewhere.

**Comment by yaccos** [comment]
Sorry for my sloppy coding style, I have now taken care of that.

**Comment by HenrikBengtsson** [comment]
Thanks.  Since this type of code has to be added to a lot of functions, I got a feeling that we might end up introducing a generic C macro for this, maybe something where one passes `norows` and `nocols` to the macro.  Let's see what happens.

**Comment by yaccos** [comment]
I also imagine that we can check user-supplied indicies for NA-values in `validateIndicies()` and generalize the solution to have flags such as `rowsHaveNA` and `colsHaveNA`. In this case `rows == NULL` should result in `rowsHaveNA == 0`.
